src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,243 @@
+                [ClassificationTypeNames.XmlLiteralCDataSection] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralComment] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralDelimiter] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralEmbeddedExpression] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralEntityReference] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralName] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralProcessingInstruction] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+                [ClassificationTypeNames.XmlLiteralText] = LSP.SemanticTokenTypes.Comment, // TO-DO: Potentially change to custom type
+            };
+
+        /// <summary>
+        /// Returns the semantic tokens data for a given document with an optional range.
+        /// </summary>
+        internal static async Task<int[]> ComputeSemanticTokensDataAsync(
+            LSP.TextDocumentIdentifier textDocument,
+            string? clientName,
+            ILspSolutionProvider solutionProvider,
+            Dictionary<string, int> tokenTypesToIndex,
+            LSP.Range? range,
+            CancellationToken cancellationToken)
+        {
+            var document = solutionProvider.GetDocument(textDocument, clientName);
+            Contract.ThrowIfNull(document);
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // By default we calculate the tokens for the full document span, although the user 
+            // can pass in a range if they wish.
+            var textSpan = range == null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
+
+            var classifiedSpans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(classifiedSpans);
+
+            // TO-DO: We should implement support for streaming once this LSP bug is fixed:
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            return ComputeTokens(text.Lines, classifiedSpans.ToArray(), tokenTypesToIndex);
+        }
+
+        private static int[] ComputeTokens(
+            TextLineCollection lines,
+            ClassifiedSpan[] classifiedSpans,
+            Dictionary<string, int> tokenTypesToIndex)
+        {
+            using var _ = ArrayBuilder<int>.GetInstance(out var data);
```

Comment:
```suggestion
            using var _ = ArrayBuilder<int>.GetInstance(classifiedSpans.Count, out var data);
```

this will help when there are a lot of spans so we don't need a lot of resizes.